code/datums/EPv2.dm

```diff
@@ -0,0 +1,142 @@
+You'll want to override receive_exonet_message() on your atom, and define what will occur when the message is received.
+The receiving atom will receive the origin atom (the atom that sent the message), the origin address, and finally the message itself.
+It's suggested to start with an if or switch statement for the message, to determine what to do.
+*/
+
+
+GLOBAL_LIST_EMPTY(all_exonet_connections)
+/datum/exonet_protocol
+	var/address = "" //Resembles IPv6, but with only five 'groups', e.g. XXXX:XXXX:XXXX:XXXX:XXXX
+	var/atom/movable/holder
+
+/datum/exonet_protocol/New(var/atom/H)
+	holder = H
+	..()
+
+
+// Proc: make_address()
+// Parameters: 1 (string - used to make into a hash that will be part of the new address)
+// Description: Allocates a new address based on the string supplied.  It results in consistant addresses for each round assuming it is not already taken..
+/datum/exonet_protocol/proc/make_address(var/string)
+	if(!string)
+		return
+	var/new_address
+	while(new_address == find_address(new_address)) //Collision test.
+		var/hash = md5(string)
+		var/raw_address = copytext(hash,1,25)
+		var/addr_0 = "fc00" //Used for unique local address in real-life IPv6.
+		var/addr_1 = hexadecimal_to_EPv2(raw_address)
+		new_address = "[addr_0]:[addr_1]"
+		string = "[string]0" //If we did get a collision, this should make the next attempt not have one.
+	address = new_address
+	GLOB.all_exonet_connections |= src
+
+
+// Proc: make_arbitrary_address()
+// Parameters: 1 (new_address - the desired address)
+// Description: Allocates that specific address, if it is available.
+/datum/exonet_protocol/proc/make_arbitrary_address(var/new_address)
+	if(new_address)
+		if(new_address == find_address(new_address) )	//Collision test.
+			return FALSE
+		address = new_address
+		GLOB.all_exonet_connections += src
+		return TRUE
+
+// Proc: hexadecimal_to_EPv2()
+// Parameters: 1 (hex - a string of hexadecimals to convert)
+// Description: Helper proc to add colons to a string in the right places.
+/proc/hexadecimal_to_EPv2(var/hex)
+	if(!hex)
+		return null
+	var/addr_1 = copytext(hex,1,5)
+	var/addr_2 = copytext(hex,5,9)
+	var/addr_3 = copytext(hex,9,13)
+	var/addr_4 = copytext(hex,13,17)
+	var/new_address = "[addr_1]:[addr_2]:[addr_3]:[addr_4]"
+	return new_address
+
+
+// Proc: remove_address()
+// Parameters: None
+// Description: Deallocates the address, freeing it for use.
+/datum/exonet_protocol/proc/remove_address()
+	address = ""
+	GLOB.all_exonet_connections -= src
+
+
+// Proc: find_address()
+// Parameters: 1 (target_address - the desired address to find)
+// Description: Searches the global list GLOB.all_exonet_connections for a specific address, and returns it if found, otherwise returns null.
+/datum/exonet_protocol/proc/find_address(var/target_address)
+	for(var/datum/exonet_protocol/exonet in GLOB.all_exonet_connections)
+		if(exonet.address == target_address)
+			return exonet.address
```

Comment:
I think you mean `return exonet` unless you want the input and output to be identical strings for some reason.

(you don't)